Add requirements and requirements matrix (#143)
w3c/lws-ucs: spec/requirements.md

>  1. <dfn>Globally Unique Identifiers</dfn> — Resources, including Entities and Storages,  shall be uniquely identifiable globally. No two distinct Resources, shall share the same identifier. A Resource may be identifiable via multiple, distinct identifiers. A Resource Owners shall be able to prove control of owned Resource Identifiers.

Reviewer: > A Resource Owners shall be able to prove control of owned Resource Identifiers.

I think this could/should be a separate requirement.

Reviewer 2: Yes. I also suggest —
```suggestion
 1. <dfn>Globally Unique Identifiers</dfn> — Resources, including Entities and Storages, shall be uniquely identifiable globally. No two distinct Resources shall share the same Identifier (though a "collective" Resource with one Identifier may be comprised of several Resources, each with its own Identifier, and actions taken on the collective Resource can affect all the Resources which comprise the collective Resource). A Resource may be identifiable via multiple, distinct identifiers. A Resource Owner shall be able to prove control of owned Resource Identifiers.
```

>  2. <dfn>Use of Service Providers</dfn> — The system shall provider a mechanism for Entities to delegate some functions to trusted Service Providers. Some interactions may require a trust relationship between Service Providers and Entities. In general, this should not impede the ability of an Entity to operate or self-host a service. Also, trust relationship are not transitive, in the sense that if an Entity trusts a Service Provider, e.g. an Identity Provider, it does not follow that another Service Provider the Entity interacts with, e.g. a Storage Provider, is under any obligation to trust said Identity Provider.

Reviewer: I find the name of this requirement "**Use** of Service Providers" a bit abstract, and when I read it, I don't recognized what is in the text (which, regardless, is a very good requirement).

```suggestion
 2. <dfn>Use of Service Providers</dfn> — The protocol shall provide a mechanism for Entities to delegate some functions to trusted Service Providers. Some interactions may require a trust relationship between Service Providers and Entities. In general, this should not impede the ability of an Entity to operate or self-host a service. Also, trust relationship are not transitive, in the sense that if an Entity trusts a Service Provider, e.g. an Identity Provider, it does not follow that another Service Provider the Entity interacts with, e.g. a Storage Provider, is under any obligation to trust said Identity Provider.
```
I suggest to replace "system" with "protocol", because this is what we are chartered to specify.

>  3. <dfn>Ownership of Storages</dfn> — An Entity shall be able to control one or more Storages across one or more Storage Providers. A Storage shall have exactly one Owner.

Reviewer: ```suggestion
 3. <dfn>Control of Storages</dfn> — An Entity shall be able to control one or more Storages across one or more Storage Providers. A Storage shall have exactly one Controller.
```
This requirement is mixing "owner" and "controller", despite the fact that we agreed to distinguish the two. I think the right one here is "control" (even though, of course, the two are related).

>  4. <dfn>Delegation of Control</dfn> — The system shall allow for an Entity to delegate control of a Storage to another Entity. Delegation could be temporary, i.e. have an expiration time, or not. An Entity shall be able to modify delegation at a later time, either by changing expiration of revoking it altogether.

Reviewer: ```suggestion
 4. <dfn>Delegation of Control</dfn> — The protocol shall allow for an Entity to delegate control of a Storage to another Entity. Delegation could be temporary, i.e. have an expiration time, or not. The first Entity shall be able to modify delegation at a later time, either by changing the expiration time of revoking it altogether.
```

Reviewer: Is this requirement only about a whole Storage, or to we also want to support delegation on a specific resource?

>  5. <dfn>Transfer of Control</dfn> — The system shall allow for an Entity to transfer, i.e. reassign, ownership of a Storage to another Entity. Transfer is final and irrevocable.

Reviewer: ```suggestion
 5. <dfn>Transfer of Control</dfn> — The protocol shall allow for an Entity to transfer, i.e. irrevocably reassign control of a Storage to another Entity.
```
I didn't like the spin of the last sentence, which sounds more like a UI warning message than a requirement... A less drastic change would be 
```suggestion
 5. <dfn>Transfer of Control</dfn> — The protocol shall allow for an Entity to transfer, i.e. reassign, ownership of a Storage to another Entity. Transfer shall be irrevocable.
```

>  6. <dfn>Storage Portability</dfn> — The system shall allow for an Entity to port, i.e. move or transfer, the entire contents of a Storage from one Storage Provider to another. Once the move is complete, the previous Storage shall be freed from any responsibility related to the Storage.

Reviewer: ```suggestion
 6. <dfn>Storage Portability</dfn> — The protocol shall allow for an Entity to port, i.e. move or transfer, the entire contents of a Storage from one Storage Provider to another. Once the move is complete, the previous Storage Provider shall be freed from any responsibility related to the Storage.
```

>  7. <dfn>Adding, Updating, Deleting Resources in Storage</dfn> — The system shall allow for Resources to be added, updated and/or deleted within a Storage by authorized Entities. In general the system shall allow for any type of resource to be stored in a Storage, however Storage Providers may impose certain limitations, such as of type or size.

Reviewer: ```suggestion
 7. <dfn>Adding, Updating, Deleting Resources in Storage</dfn> — The protocol shall allow for Resources to be added, updated and/or deleted within a Storage by authorized Entities. In general the protocol shall allow for any type of resource to be stored in a Storage, however Storage Providers may impose certain limitations, such as of type or size.
```

Reviewer: Also, I would suggest to move this requirement at the top of the list. Of course, the order does not convey any meaning, but from a reader's perspective, this one is pretty core, IMO.

>  8. <dfn>Data Sharing</dfn> — The system shall support an Entity to allow access of another Entity to a Resource it controls. In other words, a Storage Provider shall be able an Entity to access a Resource it does not control based on a verifiable proof of consent supplied by the Owner Entity.

Reviewer: The wording of this one is pretty ambiguous, and (I think) grammatically incorrect in places.
```suggestion
 8. <dfn>Data Sharing</dfn> — The protocol shall allow an Entity to grant give access to a Resource it controls to another Entity. In other words, the first Entity may allow the other Entity to perform some operations (read, modify, remove...) on the Resource. Access grant could be temporary, i.e. have an expiration time, or not. The first Entity shall be able to modify delegation at a later time, either by changing the expiration time of revoking it altogether. 
```
I reused the language used for the "delegation" requirement.

Also, I removed the mention to "proof of consent". As someone pointed out during the last meeting, some systems merely grant access based on the identity of the grantee.

>  9. <dfn>Auditable trail</dfn> — The system shall provide an auditable log of all access requests and grants to Resources and all read/write/delete of Resources.

Reviewer: ```suggestion
 9. <dfn>Auditable trail</dfn> — The protocol shall make it possible for an authorized Entity to access an auditable log of all access requests and grants to Resources and all read/write/delete of Resources.
```
> all read/write/delete of Resources

* why not create? :stuck_out_tongue: 
* do we really want all implementations to log all access to all resources?? That sounds like a strong burden on implementers / providers...

Reviewer: Also, this requirement talks about access request, but there is no requirement about requesting access...
